pennylane/templates/subroutines/all_singles_doubles.py

```diff
@@ -0,0 +1,204 @@
+                qml.templates.AllSinglesDoubles(weights, wires, hf_state, singles, doubles)
+                return qml.expval(qml.PauliZ(0))
+
+            # Evaluate the QNode for a given set of parameters
+            params = np.random.normal(0, np.pi, len(singles) + len(doubles))
+            circuit(params, hf_state, singles=singles, doubles=doubles)
+    """
+
+    num_params = 1
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, weights, wires, hf_state, singles=None, doubles=None, do_queue=True):
+
+        if len(wires) < 2:
+            raise ValueError(
+                "The number of qubits (wires) can not be less than 2; got len(wires) = {}".format(
+                    len(wires)
+                )
+            )
+
+        if not singles and not doubles:
+            raise ValueError(
+                "'singles' and 'doubles' lists can not be both empty; got singles = {}, doubles = {}".format(
+                    singles, doubles
+                )
+            )
+
+        if doubles is not None:
+            for d_wires in doubles:
+                if len(d_wires) != 4:
+                    raise ValueError(
+                        "Expected entries of 'doubles' to be of size 4; got {} of length {}".format(
+                            d_wires, len(d_wires)
+                        )
+                    )
+
+        if singles is not None:
+            for s_wires in singles:
+                if len(s_wires) != 2:
+                    raise ValueError(
+                        "Expected entries of 'singles' to be of size 2; got {} of length {}".format(
+                            s_wires, len(s_wires)
+                        )
+                    )
+
+        weights_shape = qml.math.shape(weights)
+        exp_shape = self.shape(singles, doubles)
+        if weights_shape != exp_shape:
+            raise ValueError(f"'weights' tensor must be of shape {exp_shape}; got {weights_shape}.")
+
+        # we can extract the numpy representation here since hf_state can never be differentiable
+        self.hf_state = qml.math.toarray(hf_state)
+        self.singles = singles
+        self.doubles = doubles
+
+        if hf_state.dtype != np.dtype("int"):
+            raise ValueError(f"Elements of 'hf_state' must be integers; got {hf_state.dtype}")
+
+        super().__init__(weights, wires=wires, do_queue=do_queue)
+
+    def expand(self):
+
+        weights = self.parameters[0]
+
+        with qml.tape.QuantumTape() as tape:
+
+            BasisState(self.hf_state, wires=self.wires)
+
+            for i, d_wires in enumerate(self.doubles):
+                qml.DoubleExcitation(weights[len(self.singles) + i], wires=d_wires)
+
+            for j, s_wires in enumerate(self.singles):
+                qml.SingleExcitation(weights[j], wires=s_wires)
+
+        return tape
+
+    @staticmethod
+    def shape(singles, doubles):
+        r"""Returns the shape of the tensor containing the circuit parameters.
```

Comment:
```suggestion
        r"""Returns the expected shape of the tensor that contains the circuit parameters.
```

(Maybe the wording in other templates could also be improved) :)